dhall/src/Dhall.hs

```diff
@@ -157,12 +165,47 @@ import qualified Dhall.Util
     This exception indicates that an invalid `Type` was provided to the `input`
     function
 -}
+type Extractor s a = Validation (ExtractErrors s a)
+
+newtype ExtractErrors s a =
+    ExtractErrors
+  { getErrors :: NonEmpty (ExtractError s a)
+  } deriving Semigroup
+
+typeError :: Expr s a -> Expr s a -> Extractor s a b
+typeError expected actual = Failure . ExtractErrors . (:| []) . TypeMismatch $ InvalidType expected actual
+
+extractError :: Text -> Extractor s a b
+extractError = Failure . ExtractErrors . (:| []) . ExtractError
+
+-- | opt-in monadic composition for `Extractor`. This is done this way to
+-- keep applicative behaviour on @ExtractErrors@
+validate :: Extractor s a b -> (b -> Extractor s a c) -> Extractor s a c
```

Comment:
If we don't use `validate` internally I would omit it from the API, especially given the fact that we expose the `Extractor` type's internals

Reply:
I think it's fine if people use this for non-type-errors.

I think my main issue with this function is that it's basically a `Monad` instance but without the instance.  I think if we firmly believe that the user will need `Monad` operations then we should go all the way and provide a `newtype` that exposes a `Monad` instance rather than a named bind function, because the latter will not work with `do` notation, monad transformers, or `Control.Monad` utilities.

Reply:
Ok, I'll make a monad instance with a newtype then.

Reply:
Now that I think of it, `Extractor s a` is just a type alias for `Validation (NonEmpty (ExtractError s a)) b`. I can add a new type alias for `Either (NonEmpty (ExtractError s a)) b` (named eg `MonadicExtractError`), and maybe provide a pair of  `from/to` functions (aliases of `eitherToValidation` and `validationToEither`). Or we just let implementors use `Data.Either.Validation` functions themselves